ccc-server/src/main.rs

```diff
@@ -62,6 +64,15 @@ async fn error_handler(error: BoxError) -> impl IntoResponse {
 	}
 }
 
+fn user_agent_layer() -> SetRequestHeaderLayer<HeaderValue> {
+	const USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));
```

Comment:
This will be baked into the binary, but this function is not `const`. Is `SetRequestHeaderLayer::overriding` const as well? Do we want it to be? (I think we could upstream that, even!)

Since it's not right now, the reason I ask is because it would be nice to be able to add a comment at runtime by specifying a profile variable. We should only include that if set, otherwise default to the normal `product / version` format with no stray whitespace. Note that `vars_os` has platform-specific behavior, so we might need to eat the cost of unicode validation (arguably a good idea regardless).

Reply:
[`SetRequestHeaderLayer::overriding`](https://docs.rs/tower-http/0.4.3/tower_http/set_header/struct.SetRequestHeaderLayer.html#method.overriding) (and [source](https://docs.rs/tower-http/0.4.3/src/tower_http/set_header/request.rs.html#118-120)) does not look to be const. Do we want upstream to be const? I do like the profile variable idea.

Reply:
Right… if it _does_ become `const` then we can build the entire header at compile-time. I would like to do that, but it's not a blocker for this PR.

The profile variable idea, however, will make it so that we can't use this as a const fn anymore. That's because retrieving a profile variable _has_ to be done at runtime.

Reply:
Cool, that distinction makes sense.

Reply:
Exploring these two options:

### 1. Making upstream `const`

I am unsure of how many additional changes are needed upstream around this. There is a major rewrite happening with the internals of `tower-http` so it's either a really good time or a really bad time to bring this up with that group.

```diff
--- a/tower-http/src/set_header/request.rs
+++ b/tower-http/src/set_header/request.rs
@@ -115,7 +115,7 @@ impl<M> SetRequestHeaderLayer<M> {
     ///
     /// If a previous value exists for the same header, it is removed and replaced with the new
     /// header value.
-    pub fn overriding(header_name: HeaderName, make: M) -> Self {
+    pub const fn overriding(header_name: HeaderName, make: M) -> Self {
         Self::new(header_name, make, InsertHeaderMode::Override)
     }

@@ -134,7 +134,7 @@ impl<M> SetRequestHeaderLayer<M> {
         Self::new(header_name, make, InsertHeaderMode::IfNotPresent)
     }

-    fn new(header_name: HeaderName, make: M, mode: InsertHeaderMode) -> Self {
+    const fn new(header_name: HeaderName, make: M, mode: InsertHeaderMode) -> Self {
         Self {
             make,
             header_name,
```

### 2. Profile variable

Am I following this trail correctly?

cargo.toml
```toml
[features]
user_agent = []
```

Invocation

```sh
# custom: sets "?"
cargo run --bin ccc-server  --features user_agent
# default: sets "ccc-server / 0.1.0"
cargo run --bin ccc-server
```

Implementation
```rs
fn user_agent_layer() -> SetRequestHeaderLayer<HeaderValue> {
    #[cfg(feature = "user_agent")]
    const USER_AGENT: &str = "?"; // where should this come from?
    #[cfg(not(feature = "user_agent"))]
    const USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));

    SetRequestHeaderLayer::overriding(
        HeaderName::from_static("user-agent"),
        HeaderValue::from_static(USER_AGENT),
    )
}
```

Reply:
2 is kinda what I was after. I wasn't asking for a `user_agent` feature, though — I was moreso thinking of using `std::env` to retrieve an environment variable and use that as the override. This would allow us to build a container image and adjust its behavior without rebuilding, which is useful for this option.

Reply:
Appreciate the clarification, I can adjust implementation to match.